Spurious "UpgradeResourceState" call during "destroy" workflow

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to Terraform Cloud/Enterprise, please contact tf-cloud@hashicorp.support.
If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://registry.terraform.io/browse/providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v1.1.4
on darwin_arm64
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
# any configuration
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->
*Apply*
```json
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.274317+01:00","timestamp":"2022-02-01T23:55:17.274+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.281278+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.281336+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin1956709388","network":"unix","timestamp":"2022-02-01T23:55:17.281+0100"}
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.369469+01:00","timestamp":"2022-02-01T23:55:17.369+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.376751+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.376810+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin2990233185","network":"unix","timestamp":"2022-02-01T23:55:17.376+0100"}
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.489638+01:00","timestamp":"2022-02-01T23:55:17.489+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.496614+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.496762+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin2237512697","network":"unix","timestamp":"2022-02-01T23:55:17.496+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [DEBUG] Using kubeconfig: /Users/alex/.kube/config","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.561074+01:00"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [WARN] Invalid log level: \"JSON\". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.561927+01:00"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [DEBUG] Enabling HTTP requests/responses tracing","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.561937+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.561+0100 [INFO]  [ConfigureProvider] start: EXTRA_VALUE_AT_END=93230","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.561941+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.562+0100 [INFO]  [ConfigureProvider] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.562280+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.563+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93230","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.563289+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.564+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.564658+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.823+0100 [INFO]  [PlanResourceChange] start","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.823798+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.836+0100 [INFO]  [PlanResourceChange] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.836079+01:00"}
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.894404+01:00","timestamp":"2022-02-01T23:55:17.894+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.901321+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.901376+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin4090640892","network":"unix","timestamp":"2022-02-01T23:55:17.901+0100"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [DEBUG] Using kubeconfig: /Users/alex/.kube/config","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.967187+01:00"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [WARN] Invalid log level: \"JSON\". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.967928+01:00"}
{"@level":"debug","@message":"2022/02/01 23:55:17 [DEBUG] Enabling HTTP requests/responses tracing","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.967935+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.967+0100 [INFO]  [ConfigureProvider] start: EXTRA_VALUE_AT_END=93231","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.967939+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.968+0100 [INFO]  [ConfigureProvider] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.968219+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.970+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93231","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.970572+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.971+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.971449+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:17.974+0100 [INFO]  [PlanResourceChange] start","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:17.974515+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:18.223+0100 [INFO]  [PlanResourceChange] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:18.223978+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:18.225+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93231","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:18.225291+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:18.226+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:18.226320+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:18.228+0100 [INFO]  [ApplyResourceChange] start","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:18.228428+01:00"}
{"@level":"debug","@message":"2022-02-01T23:55:18.252+0100 [INFO]  [ApplyResourceChange] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:55:18.252265+01:00"}
```
*Destroy*
```json
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:49.983567+01:00","timestamp":"2022-02-01T23:52:49.983+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:49 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:49.991418+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:49.991461+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin1712067832","network":"unix","timestamp":"2022-02-01T23:52:49.991+0100"}
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.083187+01:00","timestamp":"2022-02-01T23:52:50.082+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.089985+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.090050+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin3147096802","network":"unix","timestamp":"2022-02-01T23:52:50.089+0100"}
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.204555+01:00","timestamp":"2022-02-01T23:52:50.204+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.211475+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.211519+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin2337880449","network":"unix","timestamp":"2022-02-01T23:52:50.211+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [DEBUG] Using kubeconfig: /Users/alex/.kube/config","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.277992+01:00"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [WARN] Invalid log level: \"JSON\". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.278978+01:00"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [DEBUG] Enabling HTTP requests/responses tracing","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.278983+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.278+0100 [INFO]  [ConfigureProvider] start: EXTRA_VALUE_AT_END=93018","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.278990+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.279+0100 [INFO]  [ConfigureProvider] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.279364+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.280+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93018","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.280335+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.281+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.281728+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.538+0100 [INFO]  [PlanResourceChange] start","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.538122+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.549+0100 [INFO]  [PlanResourceChange] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.549747+01:00"}

## suspect call 
{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.599902+01:00","timestamp":"2022-02-01T23:52:50.599+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.606782+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.606826+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin122694124","network":"unix","timestamp":"2022-02-01T23:52:50.606+0100"}
{"@level":"debug","@message":"2022-02-01T23:52:50.644+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93019","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.644977+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.646+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.646298+01:00"}

{"@level":"info","@message":"configuring server automatic mTLS","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.703268+01:00","timestamp":"2022-02-01T23:52:50.702+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [ERROR] Error parsing provider name \"\": Invalid provider source string: The \"source\" attribute must be in the format \"[hostname/][namespace/]name\"","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.710055+01:00"}
{"@level":"debug","@message":"plugin address","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.710087+01:00","address":"/var/folders/cy/swrcw95d5wb15c9ff6g1m3x00000gq/T/plugin3579041967","network":"unix","timestamp":"2022-02-01T23:52:50.710+0100"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [DEBUG] Using kubeconfig: /Users/alex/.kube/config","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.775492+01:00"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [WARN] Invalid log level: \"JSON\". Defaulting to level: TRACE. Valid levels are: [TRACE DEBUG INFO WARN ERROR]","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.776263+01:00"}
{"@level":"debug","@message":"2022/02/01 23:52:50 [DEBUG] Enabling HTTP requests/responses tracing","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.776268+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.776+0100 [INFO]  [ConfigureProvider] start: EXTRA_VALUE_AT_END=93020","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.776272+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.776+0100 [INFO]  [ConfigureProvider] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.776593+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.777+0100 [INFO]  [UpgradeResourceState] start: EXTRA_VALUE_AT_END=93020","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.777814+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.778+0100 [INFO]  [UpgradeResourceState] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.778729+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:50.781+0100 [INFO]  [ApplyResourceChange] start","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:50.781752+01:00"}
{"@level":"debug","@message":"2022-02-01T23:52:51.808+0100 [INFO]  [ApplyResourceChange] end","@module":"provider.terraform-provider-kubernetes","@timestamp":"2022-02-01T23:52:51.808389+01:00"}
```

### Expected Behavior
<!--
What should have happened?
-->
`UpgradeResourceState` should not be called before `ConfigureProvider`

### Actual Behavior
<!--
What actually happened?
-->

During a `destroy` workflow, a new provider process is spawed between the `plan` and `apply` stages onto which a single `UpgradeResourceState` call is made without any preceding `ConfigureProvider` calls.

If the state upgrade logic makes use of any values from the provider block, the whole `destroy` workflow will fail. 

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->

1. Add code to `UpgradeResourceState` that tries to use values from the provider block.
2. Create minimal terraform configuration using any resource.
2. Run `terraform apply` to produce some state.
3. Run `terraform destroy`

### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/internal/terraform/context_apply2_test.go
+++ b/internal/terraform/context_apply2_test.go
@@ -1042,11 +1042,38 @@ output "out" {
 	state, diags := ctx.Apply(plan, m)
 	assertNoErrors(t, diags)
 
-	// TODO: extend this to ensure the otherProvider is always properly
-	// configured during the destroy plan
+	otherProvider.ConfigureProviderCalled = false
+	otherProvider.ConfigureProviderFn = func(req providers.ConfigureProviderRequest) (resp providers.ConfigureProviderResponse) {
+		// check that our config is complete, even during a destroy plan
+		expected := cty.ObjectVal(map[string]cty.Value{
+			"local":  cty.ListVal([]cty.Value{cty.StringVal("first-ok"), cty.StringVal("second-ok")}),
+			"output": cty.ListVal([]cty.Value{cty.StringVal("first-ok"), cty.StringVal("second-ok")}),
+			"var": cty.MapVal(map[string]cty.Value{
+				"a": cty.StringVal("first"),
+				"b": cty.StringVal("second"),
+			}),
+		})
+
+		if !req.Config.RawEquals(expected) {
+			resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf(
+				`incorrect provider config:
+expected: %#v
+got:      %#v`,
+				expected, req.Config))
+		}
+
+		return resp
+	}
 
 	opts.Mode = plans.DestroyMode
+	// skip refresh so that we don't configure the provider before the destroy plan
+	opts.SkipRefresh = true
+
 	// destroy only a single instance not included in the moved statements
 	_, diags = ctx.Plan(m, state, opts)
 	assertNoErrors(t, diags)
+
+	if !otherProvider.ConfigureProviderCalled {
+		t.Fatal("failed to configure provider during destroy plan")
+	}
 }

--- a/internal/terraform/node_provider.go
+++ b/internal/terraform/node_provider.go
@@ -37,10 +37,7 @@ func (n *NodeApplyableProvider) Execute(ctx EvalContext, op walkOperation) (diag
 	case walkValidate:
 		log.Printf("[TRACE] NodeApplyableProvider: validating configuration for %s", n.Addr)
 		return diags.Append(n.ValidateProvider(ctx, provider))
-	case walkPlan, walkApply, walkDestroy:
-		// walkPlanDestroy is purposely skipped here, since the config is not
-		// evaluated, and the provider is not needed to create delete actions
-		// for all instances.
+	case walkPlan, walkPlanDestroy, walkApply, walkDestroy:
 		log.Printf("[TRACE] NodeApplyableProvider: configuring %s", n.Addr)
 		return diags.Append(n.ConfigureProvider(ctx, provider, false))
 	case walkImport:

--- a/internal/terraform/node_resource_abstract_test.go
+++ b/internal/terraform/node_resource_abstract_test.go
@@ -121,6 +121,9 @@ func TestNodeAbstractResource_ReadResourceInstanceState(t *testing.T) {
 			},
 		},
 	})
+	// This test does not configure the provider, but the mock provider will
+	// check that this was called and report errors.
+	mockProvider.ConfigureProviderCalled = true
 
 	tests := map[string]struct {
 		State              *states.State
@@ -158,6 +161,7 @@ func TestNodeAbstractResource_ReadResourceInstanceState(t *testing.T) {
 			ctx.StateState = test.State.SyncWrapper()
 			ctx.PathPath = addrs.RootModuleInstance
 			ctx.ProviderSchemaSchema = mockProvider.ProviderSchema()
+
 			ctx.ProviderProvider = providers.Interface(mockProvider)
 
 			got, readDiags := test.Node.readResourceInstanceState(ctx, test.Node.Addr.Resource.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance))
@@ -183,6 +187,9 @@ func TestNodeAbstractResource_ReadResourceInstanceStateDeposed(t *testing.T) {
 			},
 		},
 	})
+	// This test does not configure the provider, but the mock provider will
+	// check that this was called and report errors.
+	mockProvider.ConfigureProviderCalled = true
 
 	tests := map[string]struct {
 		State              *states.State

--- a/internal/terraform/provider_mock.go
+++ b/internal/terraform/provider_mock.go
@@ -218,6 +218,11 @@ func (p *MockProvider) UpgradeResourceState(r providers.UpgradeResourceStateRequ
 	p.Lock()
 	defer p.Unlock()
 
+	if !p.ConfigureProviderCalled {
+		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("Configure not called before UpgradeResourceState %q", r.TypeName))
+		return resp
+	}
+
 	schema, ok := p.getProviderSchema().ResourceTypes[r.TypeName]
 	if !ok {
 		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("no schema found for %q", r.TypeName))

